contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,380 @@
+}
+
+type VolumeSnapshotStatus struct {
+	// The time the snapshot was successfully created
+	// +optional
+	CreationTimestamp metav1.Time `json:"creationTimestamp" protobuf:"bytes,1,opt,name=creationTimestamp"`
+
+	// Represents the lates available observations about the volume snapshot
+	Conditions []VolumeSnapshotCondition `json:"conditions" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+type VolumeSnapshotConditionType string
+
+// These are valid conditions of a volume snapshot.
+const (
+        // VolumeSnapshotConditionCreating means the snapshot is being created but
+        // it is not cut yet.
+        VolumeSnapshotConditionCreating VolumeSnapshotConditionType = "Creating"
+        // VolumeSnapshotConditionUploading means the snapshot is cut and the application
+        // can resume accessing data if core_v1.ConditionStatus is True. It corresponds
+        // to "Uploading" in GCE PD or "Pending" in AWS and core_v1.ConditionStatus is True.
+        // This condition type is not applicable in OpenStack Cinder.
+        VolumeSnapshotConditionUploading VolumeSnapshotConditionType = "Uploading"
+        // VolumeSnapshotConditionReady is added when the snapshot has been successfully created and is ready to be used.
+        VolumeSnapshotConditionReady VolumeSnapshotConditionType = "Ready"
+        // VolumeSnapshotConditionError means an error occurred during snapshot creation.
+        VolumeSnapshotConditionError VolumeSnapshotConditionType = "Error"
+)
+
+// VolumeSnapshot Condition describes the state of a volume snapshot at a certain point.
+type VolumeSnapshotCondition struct {
+	// Type of replication controller condition.
+	Type VolumeSnapshotConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=VolumeSnapshotConditionType"`
+	// Status of the condition, one of True, False, Unknown.
+	Status core_v1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+	// The last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason" protobuf:"bytes,4,opt,name=reason"`
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string `json:"message" protobuf:"bytes,5,opt,name=message"`
+}
+
+```
+
+#### The `VolumeSnapshotData` Object
+
+```GO
+
+// +genclient=true
+// +nonNamespaced=true
+
+// VolumeSnapshotData represents the actual "on-disk" snapshot object
+type VolumeSnapshotData struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	Metadata metav1.ObjectMeta `json:"metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotDataSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```

Comment:
Do we need status for `VolumeSnapshotData `?

Reply:
Since we are not binding in the in-tree controller any more, we don't need a status for VolumeSnapshotData any more because external-snapshotter will be handling the status creation and binding.